spec.md

```diff
@@ -96,7 +96,77 @@ and eventually have it accepted as a Unicode Technical Standard (UTS).
 
 ## Data Model
 
-> _The data-only representation of a message._
+A Message provides the representation of a single message.
+It takes one of two forms,
+either as a PatternMessage or a SelectMessage.
+
+```ts
+type Message = PatternMessage | SelectMessage
+
+type MessageBody = PatternElement[]
+```
+
+A PatternMessage contains a list of PatternElement values,
+some of which are directly defined literal values,
+while others are placeholders with formatted values that depend on additional data.
+
+```ts
+// Abstract data type
+interface PatternMessage {
+  value: MessageBody
+}
+
+// Canonical JSON representation
+interface PatternMessage {
+  type: 'message'
+  value: MessageBody
+}
+```
+
+SelectMessage provides for the selection of one list of PatternElement values
+to use as the message's value when formatting,
+depending on the value of one or more Selector values.
+
+Each of the SelectMessage `cases` is defined by a key of one or more string identifiers,
+and selection between them is made according to the corresponding Selector values.
+From this it follows that a valid SelectMessage must have at least as many `select` entries
+as its highest count of string entries within the keys of its `cases`.
+The `fallback` value of a Selector is used in addition to its `value`
+when selecting one of the `cases` during formatting.
+It should match exactly one of the corresponding SelectCase `key` values.
+
+As the SelectMessage `cases` uses a map with complex keys that cannot be represented in JSON,
+its canonical JSON representation is a corresponding array of `key`, `value` pairs.
+
+```ts
+// Abstract data type
+interface SelectMessage {
+  select: Selector[]
+  cases: Map<string[], PatternMessage>
+}
+
+interface Selector {
+  value: PatternElement
+  fallback?: string
+}
+
+// Canonical JSON representation
+interface SelectMessage {
```

Comment:
Yes, that's because it's the canonical JSON representation and not the abstract data type. You requested that they be defined separately.

Reply:
> Yes, that's because it's the canonical JSON representation and not the abstract data type.

No, I requested to be removed because it is redundant, and we don't care about direct mapping to JSON.

Some programming languages already have info about type (RTTI).
And those that don't will need some kind of "type" field, but it is an implementation detail.
Serialization to/from JSON is possible, but not direct.

But that was and is not required.
This is one of the reasons we decided to go with a data model: to not be restricted by syntax.

Was a decision made 2 years ago:
https://github.com/unicode-org/message-format-wg/blob/main/guidelines/goals.md

And ignoring the goals, there isn't even one issue asking for a **direct** mapping to JSON.

Reply:
> > Yes, that's because it's the canonical JSON representation and not the abstract data type.
> 
> No, I requested to be removed because it is redundant, and we don't care about direct mapping to JSON.

To be clear, _you_ Mihai have made the assertion "I don't care about JSON." Others in the working group's discussions have repeatedly noted that having a JSON-compatible representation of the data model would either be beneficial, or even required. In the recording of our last session, such statements are made by @romulocintra and @gibson042 starting from the 1h 38min mark.

> Some programming languages already have info about type (RTTI). And those that don't will need some kind of "type" field, but it is an implementation detail. Serialization to/from JSON is possible, but not direct.
> 
> But that was and is not required. This is one of the reasons we decided to go with a data model: to not be restricted by syntax.

And this is why the data model's "abstract data type" is presented here separately from the canonical JSON representation, so that no such restrictions are imposed on data model operations.

If there is some way in which specifying the canonical JSON representation of the data model hinders work with it, that obviously needs to be taken into account. But as you yourself have agreed that "in implementation there is no difference" between these representations, no such hindrance on operations actually exists; the difference is only in the abstraction level.

So we're in a situation where:
1. You have requested that we separate the two different representations of the data model from each other, and this has now been done.
2. We share an understanding that defining a canonical JSON representation does not reduce our ability to perform operations on the abstract data type representation of the data model.
3. You have stated that you don't care about the JSON representation, while others have stated that they do care about the JSON representation.

Given the above, is the current state of affairs a sufficient compromise? Everyone gets what they want.

Reply:
It's really hard to follow along this thread, Github UI is not helping and also the number of discussions overlaps and is not easy to understand sometimes 🤯🤯🤯🤯. 

## Spec Document

 I want to mention three things.

1. This is a Draft as a state on the first line of this document
2. It's a live document that will change and be adapted according to our progress.
3. Monday's brainstorm will help clarify and avoid some of this...


## JSON representation

About the JSON canonical representation of DM, I advocate that our data model should be compatible or easily converted to JSON without causing too much trouble or limiting the design of DM. We also agreed a long time ago that is not part of our Goals, but is a must if we can achieve it without "cost".  ✅  Looks like we got it so far.

### Should we care about JSON ?
We have in our deliverables list this point "A specification for a one-to-one mapping between the data model and XLIFF."

I think we should always care about the representations of DM unless they compromise the DM itself.

Now about JSON being present on the spec document, I will use the same analogy, should we have replicated interfaces for all other possible canonical representations? (Calling canonical but maybe not the right word)

TBH I asked about duplication of interfaces days ago because personally, I don't see this at the same level of the Data Model, but I also feel that if we do an effort of having different canonical representations or serialization of this model it must be connected somehow with the base representation as a whole.

We can have annexes, examples, Abstract Operations sections for JSON where those interfaces may appear(idea from ECMA specs) or any other thing that could explain this serialization or interface. 

We can try to research how those cases are solved in other bodies of standards

At moment the actual representation works for me, and while we progress we can come up with something else.

**Conclusion:**
If the unique friction point of this merge is having this duplication of interfaces to represent JSON, I think we should let it go or give a solution or alternative, that in my POV does not include the deletion of those interfaces.

Reply:
Also would like to as @stasm  @DavidFatDavidF  @grhoten  to have a look at this PR you are already reviewers

Reply:
1. I don't know what that means
I did not ask for 2 data models.
You mean that I asked to clarify what happens between parse, functions generating runtime values, and format?
And the fact that it results in 2 data models? (or one model with different states?)

2. Yes, a canonical JSON representation does not reduce out ability to do anything
But in a spec it add un-necessary noise.
It does not belong there, and it introduces "syntax artifacts" in the model.
That is the whole point to have a model: distance ourselves from a syntax, and only put there what is needed.
This is not needed, as I explained (the whole discution about programming languages with RTTI)

3. With 3 you are misrepresenting (again) my position.

I don't care about a DIRECT representation to JSON.
And nobody asked for that.

Nobody prevents a serialization to take this
```
interface PatternMessage {
  value: MessageBody
}
```
and generate a json with type:
```
{
   type: "PatternMessage "
   value: { ... }
}
```

_"while others have stated that they do care about the JSON representation."_
Can you provide a link to an issue that says "support direct serialization to JSON"?
I did give a link to "Goals and Non-Goals", which is one of the relatively few things approved by the group. It does not mention json.

I agree that JSON is nice to have.
Direct mapping to json, if it "distorts" the model, is not nice to have.